### PR TITLE
Ordinalrefproperty

### PR DESCRIPTION
**Pull request by petersteneteg**
Like ordinal property but with a get and set callback for the value.

### CODE REVIEW

File: include/inviwo/core/properties/ordinalrefproperty.h [hunk 1]
```diff
@@ -0,0 +1,611 @@
...
+    maxValue_.setAsDefault();
+    increment_.setAsDefault();
+    return *this;
+}
+
+template <typename T>
+void OrdinalRefProperty<T>::serialize(Serializer& s) const {
+    Property::serialize(s);
+
+    minConstraint_.serialize(s, this->serializationMode_);
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/ordinalrefproperty.h:499]: value must be serialized/deserialized?

File: include/inviwo/core/properties/ordinalrefproperty.h [hunk 2]
```diff
@@ -0,0 +1,611 @@
...
+            return PropertySemantics::Text;
+        } else {
+            return PropertySemantics::Default;
+        }
+    }
+};
+
+/**
+ * \ingroup properties
+ * A property representing a Ordinal value, for example int, floats.
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/ordinalrefproperty.h]: an Ordinal
Also state its difference to OrdinalProperty

File: src/core/common/inviwocore.cpp [hunk 3]
```diff
@@ -281,61 +305,49 @@ InviwoCore::InviwoCore(InviwoApplication* app)
...
-    registerProperty<DoubleQuaternionProperty>();
+    registerProperty<FilePatternProperty>();
+    registerProperty<DirectoryProperty>();
+
+    // Register ordinal property
+    using OrdinalTypes =
+        std::tuple<float, vec2, vec3, vec4, mat2, mat3, mat4, double, dvec2, dvec3, dvec4, dmat2,
+                   dmat3, dmat4, int, ivec2, ivec3, ivec4, glm::i64, unsigned int, uvec2, uvec3,
+                   uvec4, size_t, size2_t, size3_t, size4_t, glm::fquat, glm::dquat>;
+    util::for_each_type<OrdinalTypes>{}(OrdinalReghelper{}, *this);
```
Comment (ResearchDaniel) [inline src/core/common/inviwocore.cpp:342]: Might need to update the json module for the new properties as well

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
Looks good. Might be a bug in serialization though

File: include/inviwo/core/properties/ordinalrefproperty.h (hunk 1 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/ordinalrefproperty.h:499]: no the value is kept external to the property

Comment (ResearchDaniel) [reply include/inviwo/core/properties/ordinalrefproperty.h:499]: But how will the external thingy know about serialization?

Comment (petersteneteg) [reply include/inviwo/core/properties/ordinalrefproperty.h:499]: well it is up to the external thing to persist it state. maybe the processor serialise it or maybe the thing is serialisable it self.

**Review by ResearchDaniel — DISMISSED** [review]

File: include/inviwo/core/properties/ordinalrefproperty.h [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+        set_(val);
+        modified = true;
+    }
+
+    if (modified) this->propertyModified();
+}
+
+template <typename T>
+void OrdinalRefProperty<T>::set(const Property* srcProperty) {
+    if (auto prop = dynamic_cast<const OrdinalRefProperty<T>*>(srcProperty)) {
```
Comment (ResearchDaniel) [inline include/inviwo/core/properties/ordinalrefproperty.h:410]: Is this tio restrictive? Should we make it possible to also combine this one with OrdinalProperty and vice versa?

**Review by ResearchDaniel — CHANGES_REQUESTED** [review]
We need converters to link them with OrdinalProperty?
Also forgot that they need to be registered in the webbrowser module, sorry

File: include/inviwo/core/properties/ordinalrefproperty.h (hunk 4 above)
Comment (petersteneteg) [reply include/inviwo/core/properties/ordinalrefproperty.h:410]: Added the converters

**Review by ResearchDaniel — APPROVED** [review]

**Pull request merged**